src/integer_traits.rs

```diff
@@ -0,0 +1,509 @@
+//! Implementation of standard traits like Add, Sub, etc. for the Integer struct
+
+use crate::{integer::IntegerType, value::private::ReprValue, Integer};
+use std::ops::{Add, Div, Mul, Sub};
+
+impl PartialEq for Integer {
+    fn eq(&self, other: &Self) -> bool {
+        match (self.integer_type(), other.integer_type()) {
+            (IntegerType::Fixnum(a), IntegerType::Fixnum(b)) => a.as_rb_value() == b.as_rb_value(),
+            (IntegerType::Bignum(a), IntegerType::Bignum(b)) => unsafe {
+                rb_sys::rb_big_eq(a.as_rb_value(), b.as_rb_value()) == rb_sys::Qtrue.into()
+            },
+            (IntegerType::Fixnum(a), IntegerType::Bignum(b)) => unsafe {
+                let a = rb_sys::rb_int2big(a.to_isize());
+                rb_sys::rb_big_eq(a, b.as_rb_value()) == rb_sys::Qtrue.into()
+            },
+            (IntegerType::Bignum(a), IntegerType::Fixnum(b)) => unsafe {
+                rb_sys::rb_big_eq(a.as_rb_value(), b.as_rb_value()) == rb_sys::Qtrue.into()
+            },
+        }
+    }
+}
+
+impl PartialEq<i64> for Integer {
+    fn eq(&self, other: &i64) -> bool {
+        self.eq(&Integer::from_i64(*other))
+    }
+}
+
+impl PartialEq<Integer> for i64 {
+    fn eq(&self, other: &Integer) -> bool {
+        Integer::from_i64(*self).eq(other)
+    }
+}
+
+impl PartialOrd for Integer {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        match (self.integer_type(), other.integer_type()) {
+            (IntegerType::Fixnum(a), IntegerType::Fixnum(b)) => {
+                (a.as_rb_value() as i64).partial_cmp(&(b.as_rb_value() as i64))
+            }
+            (IntegerType::Bignum(a), IntegerType::Bignum(b)) => unsafe {
+                let result = rb_sys::rb_big_cmp(a.as_rb_value(), b.as_rb_value());
```

Comment:
Could you implement `rb_big_cmp` on `RBignum` (probably just `pub(crate)` for now) and delegate to that from here.

The same goes for `rb_big_plus` for the `Add` trait and so on.

Reply:
So i've tried this and it gets a bit ugly imo. what type signature should this have? 

```
fn add(&self, other: ?) -> ??
```

One thing I tried is to have both `?` and `??` be `Integer` but then it feels weird that a method in `RBignum` returns an `Integer`. Any suggestions?

Reply:
The signature that'd best model the possibilities would probably be:

``` rust
pub fn plus<T>(self, other: T) -> Result<NumericValue, Error>
where
    T: Numeric,
{
    protect(|| unsafe {
        NumericValue::from_rb_value_unchecked(rb_big_plus(
            self.as_rb_value(),
            other.as_rb_value(),
        ))
    })
}
```

but actually that's not very helpful, because we know we've got a more constrained case here where we're only passing an integer, so it's not going to error, and we'll always get an integer back.

Maybe it's best not to implement a `RBignum::plus`, at least not for this use case.